core/app/models/spree/fulfilment_changer.rb

```diff
@@ -0,0 +1,97 @@
+module Spree
+  class FulfilmentChanger
+    include ActiveModel::Validations
+
+    validates :quantity, numericality: { greater_than: 0 }
+    validates :desired_stock_location, presence: true
+    validate  :current_shipment_not_already_shipped
+    validate  :desired_shipment_different_from_current
+    validate  :enough_stock_at_desired_location, if: :handle_stock_counts?
+
+    def initialize(params = {})
+      @current_stock_location = params[:current_stock_location]
+      @desired_stock_location = params[:desired_stock_location]
+      @current_shipment       = params[:current_shipment]
+      @desired_shipment       = params[:desired_shipment]
+      @variant                = params[:variant]
+      @quantity               = params[:quantity]
+    end
+
+    def run!
+      return false if invalid?
+
+      desired_shipment.save! if desired_shipment.new_record?
+
+      available_quantity   = [desired_shipment.stock_location.count_on_hand(variant), 0].max
+      new_on_hand_quantity = [available_quantity, quantity].min
+      unstock_quantity     = desired_shipment.stock_location.backorderable?(variant) ? quantity : new_on_hand_quantity
+
+      ActiveRecord::Base.transaction do
+        if handle_stock_counts?
+          current_on_hand_quantity = [current_shipment.inventory_units.on_hand_or_backordered.size, quantity].min
+
+          current_stock_location.restock(variant, current_on_hand_quantity, current_shipment)
+          desired_stock_location.unstock(variant, unstock_quantity, desired_shipment)
+        end
+
+        update_current_shipment_inventory_units(new_on_hand_quantity, :on_hand)
+        update_current_shipment_inventory_units(quantity - new_on_hand_quantity, :backordered)
+      end
+
+      reload_shipment_inventory_units
+
+      if current_shipment.inventory_units.length.zero?
+        current_shipment.destroy!
+      else
+        current_shipment.refresh_rates
+      end
+
+      desired_shipment.refresh_rates
+
+      desired_shipment.order.reload
+      desired_shipment.order.update_with_updater!
+
+      true
+    end
```

Comment:
Could you split it into few methods to make it easier to read what is actually done?